Quick and dirty fixes for the new 1.12 Problems!

Peda1996 opened the pull request:
Just fixed the pom.xml, removed items which where causing problems in the console and added something in the config where you can toggle, to avoid the security flaws.:
only-playercommands: false

Peda1996 committed:
Dirty fixing several Bugs to make it compatible with 1.12. Also added a quick fix, where player can't hack access to the console!

Peda1996 committed:
added playercommands to config file

Peda1996 committed:
fixed small NullpointerException when only-playercommands is missing in the config

Peda1996 committed:
anti hacking is now hopefully working

XHawk87 commented:
There are some interesting ideas here. Thanks for taking the time to work on this and share it for others. Unfortunately, I can't merge the pull request at the moment due to a few issues with the code style and the approach to the specific issue. I hope to be able to in the future, if you can make some improvements and provide some clarity on a few points that I'll make on the diff.

XHawk87 commented on pom.xml:
```diff
@@ -13,44 +13,52 @@
-            <id>bukkit-repo</id>
-            <url>http://repo.bukkit.org/content/groups/public/</url>
+            <id>jitpack.io</id>
+            <url>https://jitpack.io</url>
```
This is an interesting tool, I have not come across it before. It seems to allow using github releases as a maven repo. Is that the case? I can see that being very useful.

Peda1996 replied:
Yea that's correct 👍 .
It's quite useful ^^.

XHawk87 commented on pom.xml:
```diff
@@ -13,44 +13,52 @@
         </dependency>
         <dependency>
-            <groupId>me.xhawk87</groupId>
+            <groupId>com.github.Peda1996</groupId>
```
I think it would be better to fix the main LanguageAPI pom.xml and use that instead. If you can set up a pull request for that, I would be much obliged.

XHawk87 replied:
The main language API has been fixed, you can switch to using it now

XHawk87 commented on src/main/java/me/xhawk87/CreateYourOwnMenus/CreateYourOwnMenus.java:
```diff
@@ -48,10 +51,19 @@
+    public static FileConfiguration getConfigFile() {
+        return config;
+    }
+
```
Last time I checked this storing the FileConfiguration locally like this isn't recommended in Bukkit. It's a lot of extra code for no real benefit. Best to let Bukkit handle it and just call getConfig() method when you actually need it.

Peda1996 replied:
I have tried that, but somehow this doens't worked out as I expected. (and i needed a quick fix for this issue)

XHawk87 replied:
Maybe you just needed to call `CreateYourOwnMenus.getPlugin().getConfig()`

XHawk87 replied:
Well basically you'd want to use a reference to the plugin and not the class, since `getConfig` is not static. To do this, you'd ensure that whatever code you're using it in has access to that plugin reference. Either getting passed one in when it is instantiated, or when that method is called, or you can actually get hold of a plugin instance using the Bukkit API. 

I am purposefully avoiding using singleton patterns though, as it is considered bad practice and may break some testing utilities and debugging tools, so I'd appreciate it if you didn't add instance state to the class using statics like that.

XHawk87 commented on src/main/java/me/xhawk87/CreateYourOwnMenus/CreateYourOwnMenus.java:
```diff
@@ -229,7 +241,7 @@ public boolean reloadMenu(String menuId) {
 
                 // Register the specific-opening permission for the loaded menu
-                PluginManager mgr = getServer().getPluginManager();
+                org.bukkit.plugin.PluginManager mgr = getServer().getPluginManager();
```
Is there a purpose to adding the full namespace that I am not seeing? I don't see a conflict with an identically named class. This just appears to make the code more long-winded.

Peda1996 replied:
Lol, no there is no real purpose in adding these. Somehow my IDE (intellij) had some errors with the imports, and i doens't wanted to have a look into the whole code, and therefore i used this..

XHawk87 replied:
That's weird. I wish I knew more about IntelliJ, I'm afraid I wouldn't know why it is going wrong or how to fix it. If you can find a way around it though, that'd be great. I wouldn't really like to add in bad styling to just accommodate one malfunctioning IDE.

Peda1996 replied:
Well normally intellij is much better than other IDEs (like eclipse..), sometime (not too often) it's bugging a little.

XHawk87 commented on src/main/java/me/xhawk87/CreateYourOwnMenus/Menu.java:
```diff
@@ -621,7 +621,7 @@ private Prompt parseDynamicArgs(final StringBuilder parsedCommand,
             @Override
             public String getPromptText(ConversationContext context) {
-                return plugin.translate(player, "dynamic-arg-prompt", "Please enter {0}:", promptPart);
+                return plugin.translate(player, "dynamic-arg-prompt", "{0}", promptPart);
```
Due to the way LanguageAPI works, this only changes the default prompt prefix. I am curious what your reasoning behind this is?

Peda1996 replied:
Well don't know it was 3 am in the morning when i was writing this quick and dirty fix xD.

XHawk87 replied:
Without any metrics on usage it is hard to judge whether it would benefit the greatest number of users or not. It probably works without any extra text though, so I am inclined to agree. It should probably be part of a separate PR though.

XHawk87 commented on src/main/java/me/xhawk87/CreateYourOwnMenus/commands/menu/MenuGrabCommand.java:
```diff
@@ -4,11 +4,8 @@
-import java.util.HashMap;
-import java.util.List;
-import java.util.Set;
+import java.util.*;
```
Good call on the mass import. 5 was getting a bit much 👍

XHawk87 commented on src/main/java/me/xhawk87/CreateYourOwnMenus/commands/menu/MenuGrabCommand.java:
```diff
@@ -27,21 +24,10 @@
+    private static final Set<Material> boots = new HashSet<>();
+    private static final Set<Material> leggings = new HashSet<>();
+    private static final Set<Material> chestplates = new HashSet<>();
+    private static final Set<Material> helmets = new HashSet<>();
```
You may have introduced a bug here. If you don't specify anywhere what materials are what armour it will no longer be able to recognise them.

Peda1996 replied:
Ah, this was just a quick fix, to get the plugin to work. And I also couldn't find out what these materials where for..

XHawk87 replied:
Okay, so the purpose of the /menu grab command is to essentially mapping items directly from a menu into a player's inventory. People use it for creating kits and toolbars etc.

I wanted to also include copying into the armour slots, so it needs to be able to recognise what items can actually go into those slots or introduce invalid states like cobblestone heads etc. 

I don't know why you'd be having a problem with EnumSet, it should work afaik. If you can't get it to work like that, you'd just need to set up a `static {...}` section to populate those HashSets on class creation.

It would be nice to be able to include support for the new off-hand slot too, but that should be a separate PR really.

XHawk87 commented on src/main/java/me/xhawk87/CreateYourOwnMenus/utils/MenuScriptUtils.java:
```diff
@@ -105,16 +108,23 @@ public static boolean isValidMenuItem(ItemStack item) {
+                    boolean allowServerCommands = pc != null? Boolean.parseBoolean(pc):false;
+                    if(loreString.startsWith(playerCommand)){
+                        isValid = true;
+                    }
```
I don't think this is an idea solution to the issue of `/{}` bypassing command blacklists and whitelists. `@p/` is deprecated for a reason. Players found it very unintuitive and it makes code look ugly. The command blacklists and whitelists with permission overrides are a better way to go, if we can make those work with dynamic args too. We could optionally also provide a warning to server ops when they set a non-op to creative mode with no whitelist/blacklist set up etc

XHawk87 replied:
I suppose we could look through the menu files and automatically whitelist them. 

I am not sure what problem you're having with the blacklist and whitelist though?

XHawk87 commented on src/main/resources/config.yml:
```diff
@@ -7,4 +7,8 @@
 blacklist-commands:
-  - op
\ No newline at end of file
+  - op
```
It might be a good idea to add commands that don't work in Minecraft functions to the blacklist by default, for better initial security. No banning/kicking etc

XHawk87 replied:
You're suggesting including a default whitelist as well as a blacklist? That is an interesting idea. 

Part of CYOMs goal though is to empower the user, if we start with the training wheels on, so to speak, it might give users the wrong impression about how useful it is and how much you can do with it. 

It would also mean that they can't initially see that certain menu items won't work for normal players, only for ops, not until they think they have finished and given the items to players to use. 

Although I do like the idea of adding support for command fragments e.g. `sudo @p` as opposed to just base commands `sudo` in the whitelist and blacklist. It would make it much more versatile. Although we'd have to find a way to make it work with the override permissions as well.

XHawk87 commented on src/main/java/me/xhawk87/CreateYourOwnMenus/commands/menu/MenuCopyCommand.java:
```diff
@@ -72,7 +72,7 @@ public boolean onCommand(CommandSender sender, Command command, String label, St
         for (int i = 0; i < contents.length; i++) {
             ItemStack item = contents[i];
-            if (item != null && item.getTypeId() != 0) {
+            if (item != null && item.getType().getId() != 0) {
```
The purpose of these checks is to make sure we're not trying to copy empty inventory slots etc. I suspect Bukkit want us to use materials over magic number IDs now, so if we're going to change it, it should probably be to the recommended way of doing things.

XHawk87 commented on src/main/java/me/xhawk87/CreateYourOwnMenus/listeners/MenuListener.java:
```diff
@@ -302,7 +302,7 @@ public void run() {
                 }
                 if (dosound) {
-                    player.playSound(item.getLocation(), Sound.ITEM_PICKUP, 0.2f, ((random.nextFloat() - random.nextFloat()) * 0.7F + 1.0F) * 2.0F);
+                    player.playSound(item.getLocation(), Sound.ENTITY_ITEM_PICKUP, 0.2f, ((random.nextFloat() - random.nextFloat()) * 0.7F + 1.0F) * 2.0F);
```
I believe this will break backwards compatibility. People will no longer be able to run this in earlier versions of Minecraft. There might be a way around this, but initially we'd probably just have to make a note of that on the build so that people know not to use it in earlier versions.

LiruMouse committed:
Restore functionality to is<Clothing> methods

I'm not completely confident on SKULL_ITEM being PLAYER_HEAD, but it's pretty close,
the alternative might be to list everything The Flattening changed it into...

LiruMouse committed:
Fix the broken logic of only-playercommands

Now my menus can open, and yours can not open... if that's what you want.

LiruMouse committed:
Wait, on second thought, just do this...

That logic was so rough to follow, I apparently didn't clean it up right.
Have some nice optimizations, too.

Peda1996 committed:
Merge pull request #1 from Lirusaito/master

Fix broken things

Peda1996 committed:
Added Placeholderapi Support and removed <Clothing> methods, due to they where breaking the plugin

Peda1996 committed:
Placeholders should work now, also changed item.clone. Therefore the placeholders shouldn't be replaced when using the save() function.

Peda1996 committed:
When someone is editing the Menu, placeholders are now not replaced anymore

Peda1996 committed:
Fix: Placeholder Items can now be changed
Added: Placeholders are now reloaded when edited a menu

XHawk87 commented:
Hi @Peda1996 

It's great that you're continuing to work on this, thanks for your continued support and effort. 

If you want these new features to be integrated into the main project and help a larger number of users, I recommend splitting different features into separate pull requests so that they can be worked on individually. They'll need to be brought up to the same coding standards as the rest of the project before I can accept the pull request, otherwise it will harm the overall maintainability of the project in the longer term. If they are all lumped together like this, they'll all need to reach the desired standard before any of it can be added. I really hope that you'll consider doing this, as I think it could help a lot of people.

Peda1996 committed:
remove getConfig Method,due to static access is used now

Peda1996 committed:
org.bukkit.plugin.* style bug was fixed

Peda1996 commented:
Well, due to the case that the work is mostly done, i think it would take more work to split this things up, therefore, it may be easyer to clean this code up.

For my part i think:
* 1.12 compatiblity is messy, fix ids and so on?
* PlaceholderApi is working qutite smooth, and in my opinion this code is also clean (change my mind, if im wrong ^^)
* MenuGrabber, well really don't know much about it, never used it, maybe another maintainer for this part would be nice
* the onlyplayercommands, is only a fix that server owners are able to use this plugin for now safely on theire servers. (In menues other commands are allowed tho). For the future I think this should be handled in a blacklist / whitelist manner.

XHawk87 commented on src/main/java/me/xhawk87/CreateYourOwnMenus/CreateYourOwnMenus.java:
```diff
@@ -35,9 +40,9 @@
+ * @author XHawk87, Peda1996
  */
-public class CreateYourOwnMenus extends JavaPlugin {
+public class CreateYourOwnMenus extends JavaPlugin implements Listener{
```
Generally, you'd only turn the main plugin class into a listener if it is a very small plugin and everything is in that one class. We've already moved out to having external listener classes, so for the sake of consistency and knowing where to look for them, they should all be kept together in the `me.xhawk87.CreateYourOwnMenus.listeners` package.

Peda1996 replied:
Oh, yea thanks for the info,
This was done by me by mistake..
Listener can be removed without problems

XHawk87 commented on src/main/java/me/xhawk87/CreateYourOwnMenus/CreateYourOwnMenus.java:
```diff
@@ -48,6 +53,7 @@
     private Map<String, ScriptCommand> scriptCommands = new HashMap<>();
     private LanguageWrapper language;
 
+
```
Surplus line

XHawk87 commented on src/main/java/me/xhawk87/CreateYourOwnMenus/CreateYourOwnMenus.java:
```diff
@@ -89,6 +95,9 @@ public void run() {
         language = new LanguageWrapper(this, "eng");
+
+
+
```
You have some surplus lines here

XHawk87 commented on src/main/java/me/xhawk87/CreateYourOwnMenus/Menu.java:
```diff
@@ -99,23 +97,72 @@ public Inventory getInventory() {
+            inventory.setItem(p.slot, item);
+        }
+    }
+
```
I think there is enough here to do with placeholders that they deserve to have their own package. This separation will also help to serve as a reminder that these are two separate systems, and CYOM needs to be able to work without PlaceholderAPI being present. 

Btw, did you test to ensure that CYOM still works without PlaceholderAPI?

Peda1996 replied:
(restoreInvPlaceholders) This method is needed to recover the inventory, e.g. when an user opens a menu, where %server_online% is used, it gets replaced to the player count e.g. 10. When now this method wasn't excisting, the placeholder would be lost, due to the case, that the item in the inventory has the value 10 saved. 

Also CYOM is still working without Placeholderapi, this line ensures this:
`        if (!Bukkit.getPluginManager().isPluginEnabled("PlaceholderAPI"))`
So only if the plugin is there, the placeholderlist remains empty, and will never be used!

Descriptions of the methods will be added soon

XHawk87 commented on pom.xml:
```diff
@@ -13,44 +13,69 @@
-  
+
+
+
```
Surplus lines

Peda1996 replied:
Never thought bad of surplus lines. But if you don't like them I can delete them ^^.

XHawk87 commented on pom.xml:
```diff
@@ -13,44 +13,69 @@
+        </dependency>
+
+
+
```
Surplus lines

XHawk87 commented on src/main/java/me/xhawk87/CreateYourOwnMenus/CreateYourOwnMenus.java:
```diff
@@ -10,6 +10,8 @@
 import java.util.Map;
 import java.util.Set;
+
+
```
Surplus line

XHawk87 commented on src/main/java/me/xhawk87/CreateYourOwnMenus/Menu.java:
```diff
@@ -99,23 +97,72 @@ public Inventory getInventory() {
     }
 
+    /**
+     * Schedules an asyncronous save of this menu, and also restores the placeholders!
```
Restores in what way? From what to what?

Peda1996 replied:
Yea, I will expand the description.
This restores the replaced items in the inventory. When the item title %server_online% was replaced with 1. This is needed to restore the 1 to %server_online%, so it can be saved to the .yml file, and used another time.

XHawk87 commented on src/main/java/me/xhawk87/CreateYourOwnMenus/Menu.java:
```diff
@@ -214,9 +261,13 @@ private void onLoad(FileConfiguration data) {
             inventory.setItem(slot, item);
+
+            //check for placeholders
+            hasItemPlaceholders(slot,item);
```
Normally a function with a name that poses a question is expected to return the answer to the question posed. `hasItemPlaceholders` you would think would return a true or false, and you use it to determine whether you should do something based upon that fact. What is this really doing and could it have a more appropriate name?

Peda1996 replied:
Well you're damn right, originally this method returned true or false, and it was also used, but later I decided to change it. But the one thing that remaind was the name..

I will change it to: addPlaceholdersToList()

XHawk87 commented on src/main/java/me/xhawk87/CreateYourOwnMenus/Menu.java:
```diff
@@ -214,9 +261,13 @@ private void onLoad(FileConfiguration data) {
         }
     }
 
+
```
Surplus line

Peda1996 replied:
I sure love this lines it seems ^^

XHawk87 commented on src/main/java/me/xhawk87/CreateYourOwnMenus/Menu.java:
```diff
@@ -253,6 +307,101 @@ private void onReload(FileConfiguration data) {
     }
 
+    /**
+     * Checks if there is a placeholder in the string
```
This function doesn't just check, it does something. That should be stated in the description, and the name should properly reflect that.

Peda1996 replied:
Originally it was only checking, but I changed the functionality, and the description remained..
I will fix this..

XHawk87 commented on src/main/java/me/xhawk87/CreateYourOwnMenus/Menu.java:
```diff
@@ -253,6 +307,101 @@ private void onReload(FileConfiguration data) {
+     * Placeholder Data and simple inner class
+     */
+
+
```
Surplus lines

Peda1996 replied:
Well, i love them xD

XHawk87 commented on src/main/java/me/xhawk87/CreateYourOwnMenus/Menu.java:
```diff
@@ -253,6 +307,101 @@ private void onReload(FileConfiguration data) {
+     */
+
+
+    private List<Placeholder> placeholderList = new ArrayList<Placeholder>();
```
Class fields should be declared at the top of the class. This is so that it doesn't waste a programmers time scrolling through the whole class, when they could get all of that info together in one place at the start.

Peda1996 replied:
Will be changed in the next commit

XHawk87 commented on src/main/java/me/xhawk87/CreateYourOwnMenus/Menu.java:
```diff
@@ -253,6 +307,101 @@ private void onReload(FileConfiguration data) {
+        private Map<Integer,String> lorePlaceholderPositions = new TreeMap<>();
+        //private List<Integer> lorePlaceholderPositions = new ArrayList<>();
+
+
```
Surplus lines

Peda1996 replied:
Well..

XHawk87 commented on src/main/java/me/xhawk87/CreateYourOwnMenus/Menu.java:
```diff
@@ -253,6 +307,101 @@ private void onReload(FileConfiguration data) {
+        if (!Bukkit.getPluginManager().isPluginEnabled("PlaceholderAPI"))
+            return;
+
+
```
Surplus line

Peda1996 replied:
Yep..

XHawk87 commented on src/main/java/me/xhawk87/CreateYourOwnMenus/Menu.java:
```diff
@@ -253,6 +307,101 @@ private void onReload(FileConfiguration data) {
+                if(PlaceholderAPI.containsPlaceholders(s)){
+                    placeholder.lorePlaceholderPositions.put(index,s);
+                    retVal = true;
+
```
Surplus line

Peda1996 replied:
ok

XHawk87 commented on src/main/java/me/xhawk87/CreateYourOwnMenus/Menu.java:
```diff
@@ -253,6 +307,101 @@ private void onReload(FileConfiguration data) {
+        if(retVal){
+            placeholderList.add(placeholder);
+        }
+
```
Surplus line

Peda1996 replied:
fix

XHawk87 commented on src/main/java/me/xhawk87/CreateYourOwnMenus/Menu.java:
```diff
@@ -253,6 +307,101 @@ private void onReload(FileConfiguration data) {
+    }
+
+    /**
+     * Placeholder Data and simple inner class
```
Inner classes like this should generally be avoided as they are less visible to other coders. It also helps to keep data manipulation and definition in the same place for easy reference. If something is manipulating something in an object, it is best done within that class, and you provide methods for others to use with good descriptions of their purpose to make it clear what is happening to the object at every stage. If you use inner classes like this, that starts to break down, as it could be manipulated from anywhere, and that makes it harder to debug.

XHawk87 commented on src/main/java/me/xhawk87/CreateYourOwnMenus/Menu.java:
```diff
@@ -253,6 +307,101 @@ private void onReload(FileConfiguration data) {
+
+    }
+
+
```
Surplus line

XHawk87 commented on src/main/java/me/xhawk87/CreateYourOwnMenus/Menu.java:
```diff
@@ -253,6 +307,101 @@ private void onReload(FileConfiguration data) {
+                for(Map.Entry<Integer, String> entry : p.lorePlaceholderPositions.entrySet()){
+                    Integer slot = entry.getKey();
+                    String lore = entry.getValue();
+                    metaLore.set(slot,PlaceholderAPI.setPlaceholders(player,lore));
```
```suggestion
                    metaLore.set(slot,PlaceholderAPI.setPlaceholders(player, lore));
```

Auto-formatting the page would probably help resolve a lot of these minor issues with newlines and whitespaces. Just to keep things neater for the next people to work on it.

XHawk87 commented:
Okay, to help with this, I've reviewed the code and annotated where I think improvements can be made.

>     * PlaceholderApi is working qutite smooth, and in my opinion this code is also clean (change my mind, if im wrong ^^)

Good work on this. There were only a few issues that I pointed out, and I just think it could do with being kept in a separate package. 

>     * MenuGrabber, well really don't know much about it, never used it, maybe another maintainer for this part would be nice

What's wrong with the `/menu grab` command?

>     * the onlyplayercommands, is only a fix that server owners are able to use this plugin for now safely on theire servers. (In menues other commands are allowed tho). For the future I think this should be handled in a blacklist / whitelist manner.

So long as this defaults to off, and we properly document and highlight this feature, I'm okay with it as it is, but it does need improvement. It basically hamstrings what you can do with CYOM on creative-mode servers. 

>     * Well I also sadly don't know how to make the plugin compatible with multiple versions

Don't worry about this. I think trying to maintain backwards compatibility is becoming too great a task. If people want the latest features of CYOM, they'll have to get the latest version of Minecraft too.

Peda1996 commented:
Im away for multiple days. So I think I can commit changes next week or so.

XHawk87 commented:
That's great. Thanks for doing this :)

Peda1996 committed:
cleaned up the Placeholders, and swapped them into extra classes!

Peda1996 committed:
Fixed a bug, and fixed some style

davidmvos commented:
Im closing this as this is a stale PR from 6 Years ago

I will be coming back to the things mentioned here tho
